consensus/src/ledger/mod.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    marker::PhantomData,
+    ops::{Deref, DerefMut},
+    path::Path,
+    sync::Arc,
+};
+
+use anyhow::*;
+use dyn_clone::DynClone;
+use smallvec::SmallVec;
+use snarkos_storage::Digest;
+use snarkvm_algorithms::{
+    merkle_tree::{MerklePath, MerkleTreeDigest},
+    CommitmentScheme,
+    MerkleParameters,
+    SignatureScheme,
+};
+use snarkvm_dpc::{
+    testnet1::{BaseDPCComponents, Transaction},
+    Block,
+    LedgerScheme,
+    TransactionScheme,
+};
+
+mod merkle;
+pub use merkle::MerkleLedger;
+mod indexed_merkle_tree;
+pub use indexed_merkle_tree::IndexedMerkleTree;
+use snarkvm_parameters::{LedgerMerkleTreeParameters, Parameter};
+use snarkvm_utilities::{FromBytes, ToBytes};
+
+pub trait Ledger: Send + Sync + DynClone {
+    fn extend(&mut self, new_commitments: &[Digest], new_serials: &[Digest], new_memos: &[Digest]) -> Result<Digest>;
+
+    fn rollback(&mut self, commitments: &[Digest], serials: &[Digest], memos: &[Digest]) -> Result<()>;
+
+    fn commitment_len(&self) -> usize;
+
+    fn contains_commitment(&self, commitment: &Digest) -> bool;
+
+    fn commitment_index(&self, commitment: &Digest) -> Option<usize>;
+
+    fn contains_serial(&self, serial: &Digest) -> bool;
+
+    fn contains_memo(&self, memo: &Digest) -> bool;
+
+    fn validate_digest(&self, digest: &Digest) -> bool;
+
+    fn digest(&self) -> Digest;
+
+    fn generate_proof(&self, commitment: &Digest, index: usize) -> Result<Vec<(Digest, Digest)>>;
+
+    /// checks if a ledgers state is consistent
+    fn validate_ledger(&self) -> bool;
+}
+
+pub struct DynLedger(pub Box<dyn Ledger>);
+
+impl Clone for DynLedger {
+    fn clone(&self) -> Self {
+        DynLedger(dyn_clone::clone_box(&*self.0))
+    }
+}
+
+impl Deref for DynLedger {
+    type Target = dyn Ledger;
+
+    fn deref(&self) -> &Self::Target {
+        &*self.0
+    }
+}
+
+impl DerefMut for DynLedger {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut *self.0
+    }
+}
+
+pub struct DeserializedLedger<'a, C: BaseDPCComponents> {
```

Comment:
Is there a `SerializedLedger`? I think this could just be `Ledger`.

Reply:
it's the inverse of our `SerialX` types. I treat something in snarkVM land as 'deserialized', otherwise it is 'serialized'. since this enters snarkvm land, it is `deserialized`.

Reply:
Some objects had a `VM` prefix, which is perhaps more clear for this context?